python/ray/util/client/server/proxier.py

```diff
@@ -0,0 +1,335 @@
+
+    def ClusterInfo(self, request,
+                    context=None) -> ray_client_pb2.ClusterInfoResponse:
+
+        # NOTE: We need to respond to the PING request here to allow the client
+        # to continue with connecting.
+        if request.type == ray_client_pb2.ClusterInfoType.PING:
+            resp = ray_client_pb2.ClusterInfoResponse(json=json.dumps({}))
+            return resp
+        return self._call_inner_function(request, context, "ClusterInfo")
+
+    def Terminate(self, req, context=None):
+        return self._call_inner_function(req, context, "Terminate")
+
+    def GetObject(self, request, context=None):
+        return self._call_inner_function(request, context, "GetObject")
+
+    def PutObject(self, request: ray_client_pb2.PutRequest,
+                  context=None) -> ray_client_pb2.PutResponse:
+        return self._call_inner_function(request, context, "PutObject")
+
+    def WaitObject(self, request, context=None) -> ray_client_pb2.WaitResponse:
+        return self._call_inner_function(request, context, "WaitObject")
+
+    def Schedule(self, task, context=None) -> ray_client_pb2.ClientTaskTicket:
+        return self._call_inner_function(task, context, "Schedule")
+
+
+def forward_streaming_requests(grpc_input_generator: Iterator[Any],
+                               output_queue: "Queue") -> None:
+    """
+    Forwards streaming requests from the grpc_input_generator into the
+    output_queue.
+    """
+    try:
+        for req in grpc_input_generator:
+            output_queue.put(req)
+    except grpc.RpcError as e:
+        logger.debug("closing dataservicer reader thread "
+                     f"grpc error reading request_iterator: {e}")
+    finally:
+        # Set the sentinel value for the output_queue
+        output_queue.put(None)
+
+
+def prepare_runtime_init_req(req: ray_client_pb2.InitRequest
+                             ) -> Tuple[ray_client_pb2.InitRequest, JobConfig]:
+    """
+    Extract JobConfig and possibly mutate InitRequest before it is passed to
+    the specific RayClient Server.
+    """
+    job_config = JobConfig()
+    if req.job_config:
+        import pickle
+        job_config = pickle.loads(req.job_config)
+
+    return (req, job_config)
+
+
+class DataServicerProxy(ray_client_pb2_grpc.RayletDataStreamerServicer):
+    def __init__(self, proxy_manager: ProxyManager):
+        self.proxy_manager = proxy_manager
+
+    def Datapath(self, request_iterator, context):
+        client_id = _get_client_id_from_context(context)
+        if client_id == "":
+            return
+        logger.debug(f"New data connection from client {client_id}: ")
+
+        init_req = next(request_iterator)
+        init_type = init_req.WhichOneof("type")
+        assert init_type == "init", ("Received initial message of type "
+                                     f"{init_type}, not 'init'.")
+
+        modified_init_req, job_config = prepare_runtime_init_req(init_req.init)
```

Comment:
```
new_job_config = prepare_runtime_init_req(job_config)
# use new job config to update the original init request
```

Reply:
@iycheng We also need to modify the `InitReq`. The reason for this is that we want the mutated one to be passed to the Specific Server so that it starts ray with `ray.init(job_config=InitReq.job_config)`